### PR TITLE
 from_pretrained: check that the pretrained model is for the right model architecture

### LINKED ISSUE #10293: [pretrained] model classes aren't checking the arch of the pretrained model it loads
While comparing different models trained on xsum (most of which are Bart) I made a mistake and passed "google/pegasus-xsum" to `BartForConditionalGeneration`  

```
BartForConditionalGeneration.from_pretrained("google/pegasus-xsum")
```

I got:

```
Some weights of the model checkpoint at google/pegasus-xsum were not used when initializing BartForConditionalGeneration: ['model.encoder.layer_norm.weight', 'model.encoder.layer_norm.bias', 'model.decoder.layer_norm.weight', 'model.decoder.layer_norm.bias']
- This IS expected if you are initializing BartForConditionalGeneration from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BartForConditionalGeneration from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BartForConditionalGeneration were not initialized from the model checkpoint at google/pegasus-xsum and are newly initialized: ['model.encoder.embed_positions.weight', 'model.encoder.layernorm_embedding.weight', 'model.encoder.layernorm_embedding.bias', 'model.decoder.embed_positions.weight', 'model.decoder.layernorm_embedding.weight', 'model.decoder.layernorm_embedding.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Traceback (most recent call last):
  File "./bart-summarize2.py", line 8, in <module>
    tokenizer = BartTokenizer.from_pretrained(mname)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/tokenization_utils_base.py", line 1788, in from_pretrained
    return cls._from_pretrained(
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/tokenization_utils_base.py", line 1860, in _from_pretrained
    tokenizer = cls(*init_inputs, **init_kwargs)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/models/roberta/tokenization_roberta.py", line 159, in __init__
    super().__init__(
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/models/gpt2/tokenization_gpt2.py", line 179, in __init__
    with open(vocab_file, encoding="utf-8") as vocab_handle:
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

Any reason why the model class doesn't check that it's being fed a wrong architecture? It could detect that and give a corresponding error message, rather than spitting random errors like above? I was pretty sure it was a bug in pegasus model until I noticed that pegasus != Bart.

Thanks.

@LysandreJik

### PR DESCRIPTION
**Pull request by vimarshc**
# What does this PR do?
Adding Checks to the from_pretrained workflow to check the model name passed belongs to the model being initiated. 
Same checks need to be added for Tokenizer. 

Fixes https://github.com/huggingface/transformers/issues/10293

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

**Comment by LysandreJik** [comment]
Hi @vimarshc, thank you for opening this PR! Could you:
- rebase your PR on the most recent master so that the failing tests don't fail anymore
- run `make fixup` at the root of your repository to fix your code quality issue (More information related to this on step 5 of [this document](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests)

**Comment by stas00** [comment]
Awesome!

Would you like to attempt to add a test for this check? 

We need to use tiny models so it's fast and I made the suggestions here:
https://github.com/huggingface/transformers/issues/10293#issuecomment-784630105

If you're not sure how to do it please let me know and I will add a test.

**Comment by vimarshc** [comment]
Hi @stas00, 
I'd like to add the tests myself if that's ok. I have to add the same checks for the `from_pretrained` for Tokenizer however it's not as straightforward. The Tokenizer's `from_pretrained` is written with some assumptions in mind and I'm not entirely sure where to add the check. Here's the `from_pretrained` method for Tokenizers. 


Regardless, I shall try to add the test for this assertion I've already added and the changes mentioned by @LysandreJik in the next 24 hours.

**Comment by stas00** [comment]
OK, so your change works for the model and the config:
```
PYTHONPATH=src python -c 'from transformers import PegasusForConditionalGeneration; PegasusForConditionalGeneration.from_pretrained("patrickvonplaten/t5-tiny-random")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/modeling_utils.py", line 975, in from_pretrained
    config, model_kwargs = cls.config_class.from_pretrained(
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/configuration_utils.py", line 387, in from_pretrained
    assert (
AssertionError: You tried to initiate a model of type 'pegasus' with a pretrained model of type 't5'
```
same for:
```
PYTHONPATH=src python -c 'from transformers import PegasusConfig; PegasusConfig.from_pretrained("patrickvonplaten/t5-tiny-random")'
```

As you discovered - and I didn't know - the tokenizer doesn't seem to need the config file, so it doesn't look there is a way to check that the tokenizer being downloaded is of the right kind. I will ask.

**Comment by vimarshc** [comment]
Hi @stas00, 
Thanks for the update. 
Will take a pull, add the test and go through the checklist before pushing the changes. 
Will try to push in a few hours.

**Comment by stas00** [comment]
I'm puzzled. why did you undo my fix?

**Comment by vimarshc** [comment]
Hi, 
Apologies. 
I rebased my branch and assumed had to force push which deleted your changes.

**Comment by stas00** [comment]
you messed up your PR branch - so this PR now contains dozens of unrelated changes.

You can do a soft reset to the last good sha, e.g.:
```
git reset --soft d70a770
git commit
git push -f
```

File: tests/test_modeling_common.py
Comment (stas00) [inline tests/test_modeling_common.py:1293]: if you were to test T5 against t5 then you don't need to add `TINY_BERT = "lysandre/tiny-bert-random"` - one model is enough here - use 2 classes instead.

File: src/transformers/configuration_utils.py
Comment (stas00) [inline src/transformers/configuration_utils.py:391]:
```suggestion
        assert (
            config_dict["model_type"] == cls.model_type
        ), f"You tried to initiate a model of type '{cls.model_type}' with a pretrained model of type '{config_dict['model_type']}'"
```
as I already tried to communicate earlier - we don't use `format`

File: tests/test_modeling_common.py
Comment (stas00) [inline tests/test_modeling_common.py:1296]:
```suggestion
        with self.assertRaises(Exception) as context:
            BertModel.from_pretrained(TINY_T5)
        self.assertTrue("You tried to initiate a model of type" in str(context.exception))
```

Let's check the actual assert message here, just in case it asserts on something else and then this test would be misleading.

Just please test that it works. thank you.

File: src/transformers/configuration_utils.py
Comment (vimarshc) [reply src/transformers/configuration_utils.py:391]: Hi @stas00, 
I used the commands from step 5 [here](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests). I ran `make style`, `make quality` and `make fixup` in that order. 

The suggestion you've made is using f-strings whereas I am using `format` (would have gone ahead with f-strings myself but when I looked around the codebase I saw `format` being used and assumed that  `format` is being used across the library

File: tests/test_modeling_common.py
Comment (vimarshc) [reply tests/test_modeling_common.py:1293]: Hi,
I don't understand the last part about using two classes. 
Should I add a separate class for this particular test?

File: src/transformers/configuration_utils.py
Comment (stas00) [reply src/transformers/configuration_utils.py]: If I understand it correctly, this library sets a standard forward but doesn't insist on spending short resources on making everything consistent. So if you look at most recent code it uses f-strings, hence the recommendation.

Comment (stas00) [reply src/transformers/configuration_utils.py]: BTW, you only need to run `make fixup` - much faster.

Comment (vimarshc) [reply src/transformers/configuration_utils.py]: Got it. Thanks for the clarification. 
I've added the change with f-strings.

Comment (stas00) [reply src/transformers/configuration_utils.py]: BTW, checked in with the group - it's f-strings - we just didn't have time to adjust old code - will document this somewhere.

File: tests/test_modeling_common.py
Comment (stas00) [reply tests/test_modeling_common.py]: The change you applied is it. I was trying to communicate that the test needs only one pretrained model and one good class and one bad class = 2 classes.

**Review by stas00 — APPROVED** [review]
Looks very good. Thank you for bearing my requests.

It Looks like this check found some bugs in our code. So we will need to resolve those before merging this. I will update you when this is done.

**Comment by stas00** [comment]
@patrickvonplaten, @patil-suraj - your help is needed here.

BlenderbotSmall has an inconsistency. It declares its model type as "blenderbot-small":
```
src/transformers/models/auto/configuration_auto.py:        ("blenderbot-small", BlenderbotSmallConfig),
src/transformers/models/auto/configuration_auto.py:        ("blenderbot-small", "BlenderbotSmall"),
src/transformers/models/blenderbot_small/configuration_blenderbot_small.py:    model_type = "blenderbot-small"
```
but the models are all "blenderbot": https://huggingface.co/facebook/blenderbot-90M/blob/main/config.json

So this new sanity check this PR is trying to add fails. 
```
        config_dict, kwargs = cls.get_config_dict(pretrained_model_name_or_path, **kwargs)
>       assert (
            config_dict["model_type"] == cls.model_type
        ), f"You tried to initiate a model of type '{cls.model_type}' with a pretrained model of type '{config_dict['model_type']}'"
E       AssertionError: You tried to initiate a model of type 'blenderbot-small' with a pretrained model of type 'blenderbot'
```

What shall we do?

**Comment by vimarshc** [comment]
Hi @stas00 
Will add the check you mentioned today.

**Comment by stas00** [comment]
Looks good, @vimarshc 

So we are down to one failing test:

```
tests/test_modeling_blenderbot_small.py::Blenderbot90MIntegrationTests::test_90_generation_from_short_input
```

**Comment by stas00** [comment]
I wonder if we could sort of cheat and do:

```
if not cls.model_type in config_dict["model_type"]: assert ...
```

so this will check whether the main type matches as a substring, of a sub-type. It's not a precise solution, but will probably catch the majority of mismatches.

Actually for t5/mt5 it's reversed. `model_type` are t5 and mt5, but both have `T5ForConditionalGeneration` as `architecture`. So I think this check would fail. Hmm, perhaps checking if one is a subset of another in either direction?

```
if not (cls.model_type in config_dict["model_type"] or config_dict["model_type"] in cls.model_type): assert ...
```

**Comment by patil-suraj** [comment]
>BlenderbotSmall has an inconsistency. It declares its model type as "blenderbot-small":

@stas00  You are right. Before the BART refactor all `blenderbot` models shared the same model class, The `model_type`  on the hub should actually `blenderbot-small`. I will fix that.

**Comment by patil-suraj** [comment]
I updated the config https://huggingface.co/facebook/blenderbot-90M/blob/main/config.json.

And actually, there's a new version of `blenderbot-90M` , https://huggingface.co/facebook/blenderbot_small-90M

It's actually the same model, but with the proper name. The blenderbot small test uses `blenderbot-90M` which should be changed to use this new model.

**Comment by vimarshc** [comment]
Hi @stas00, 
The fuzzy match approach will not work for the case 'distilbert' vs 'bert'.

**Comment by stas00** [comment]
> Hi @stas00,
> The fuzzy match approach will not work for the case 'distilbert' vs 'bert'.

That's an excellent counter-example! As I proposed that it might mostly work ;)

But it looks like your original solution will now work after @patil-suraj fixing.

**Comment by stas00** [comment]
> I updated the config https://huggingface.co/facebook/blenderbot-90M/blob/main/config.json.
> 
> And actually, there's a new version of `blenderbot-90M` , https://huggingface.co/facebook/blenderbot_small-90M
> 
> It's actually the same model, but with the proper name. The blenderbot small test uses `blenderbot-90M` which should be changed to use this new model.

Thanks Suraj!

Since it's sort of related to this PR, do you want to push the change in here, or do it in another PR?

**Comment by stas00** [comment]
Oh bummer, we have 2 more in TF land:
```
FAILED tests/test_modeling_tf_flaubert.py::TFFlaubertModelTest::test_compile_tf_model
FAILED tests/test_modeling_tf_flaubert.py::TFFlaubertModelTest::test_save_load
```
same issue for both tests:
```
E           AssertionError: You tried to initiate a model of type 'xlm' with a pretrained model of type 'flaubert'
```

@LysandreJik, who can help resolving this one? Thank you!

**Comment by LysandreJik** [comment]
Yes, I'll take a look as soon as possible!

**Comment by LysandreJik** [comment]
I fixed the tests related to FlauBERT. Flax test is a flaky test that @patrickvonplaten is working on solving, and should not block this PR.

**Review by LysandreJik — APPROVED** [review]
Overall LGTM. Would like to merge that after the v4.4.0 that comes out tomorrow so that we have time to test it on the `master` branch before putting it in a version.

**Comment by stas00** [comment]
Thank you for taking care of this, @LysandreJik 

I suppose we will take care of potentially doing the same for the Tokenizer validation in another PR.

**Comment by LysandreJik** [comment]
With the tokenizer it'll likely be a bit more complex, as it is perfectly possible to have decoupled models/tokenizers, e.g., a BERT model and a different tokenizer like it is the case in [BERTweet (config.json)](https://huggingface.co/vinai/bertweet-base/blob/main/config.json).

**Comment by stas00** [comment]
Indeed, I think this will require a change where there is a required `tokenizer_config.json` which identifies itself which arch it belongs to, so while it should be possible to mix a model and tokenizer from different architectures, this shouldn't fail with random misleading errors:

```
python -c 'from transformers import BartTokenizer; BartTokenizer.from_pretrained("prajjwal1/bert-tiny")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/tokenization_utils_base.py", line 1693, in from_pretrained
    raise EnvironmentError(msg)
OSError: Can't load tokenizer for 'prajjwal1/bert-tiny'. Make sure that:

- 'prajjwal1/bert-tiny' is a correct model identifier listed on 'https://huggingface.co/models'

- or 'prajjwal1/bert-tiny' is the correct path to a directory containing relevant tokenizer files
```

but to indicate to the user that they got either the wrong tokenizer class or the the tokenizer identifier, since the above error is invalid - it's the correct identifier 

As can be seen from:
```
python -c 'from transformers import BertTokenizer; BertTokenizer.from_pretrained("prajjwal1/bert-tiny")'
```
which works.

(and it erroneously says "model identifier" and there is no model here, but that's an unrelated minor issue).

And of course there are many other ways I have seen this mismatch to fail, usually a lot noisier when it's missing some file.

**Comment by stas00** [comment]
@LysandreJik, I rebased this PR and it looks good. v4.4.0 is out so we can probably merge this one now.

Thank you.

**Comment by LysandreJik** [comment]
Indeed, this is great! Thanks a lot @vimarshc and @stas00 for working on this.

**Comment by stas00** [comment]
So should I create a new issue for doing the same for the Tokenizers? I think it'd be much more complicated since we don't save any tokenizer data at the moment that puts the tokenizer in any category/architecture.

**Comment by vimarshc** [comment]
Hi, 
Thanks, @stas00 for providing the guidance to close this issue. This is my first contribution to transformers so you can imagine my excitement. :D 
I understand that a similar change for Tokenizer will be a bit more complicated. Would love to take a shot at fixing that as well. :)

**Comment by stas00** [comment]
I'm glad to hear it was a good experience for you, @vimarshc.

I'm not quite sure yet how to tackle the same for tokenizers. I will try to remember to tag you if we can think of an idea on how to approach this task.

**Pull request merged**
